### ISSUE TITLE
Allow overriding fields with a stricter type

### ISSUE THREAD

**Issue by cburschka** [kalessil/phpinspectionsea#152]
I believe the inspection "class overrides a field of a parent class" is a little too broad. There is a use-case for redeclaring a field with a subtype of the one used by the parent. For example, consider this code:

```php

abstract class FruitHandler {
  /**
   * @var FruitInterface
   */
  protected $fruit

  protected function form() { ... }
}

class AppleHandler {
  /**
   * @var AppleInterface
   */
  protected $fruit;

  protected function form() { ... }
}
```

By redeclaring the field, I can mark it safe to be used as the subtype in my methods, and I only need to actually ensure that in the places where it gets set. The alternative is to either make an inline variable type declaration in every method that uses the field, or to maintain *two* references to the same value ($this->fruit and $this->apple) which seems needlessly complicated.

The inspection could instead check that fields are only overridden with stricter types, which would ensure that both the parent and the extending class can safely use the field as they declared it.

**Comment by kalessil** [comment]
Thank you for reporting @cburschka.

As I understand, the intention is to re-declare the type annotation.
I think that the inspection reported at the right place because we could do following:

```php
/**
  * @property AppleInterface $fruit <description>
  */
class AppleHandler {
}
```

IMO such approach is more clean, as it does not introduce additional entities in the class inheritance tree.

**Comment by cburschka** [comment]
Thanks very much! I hadn't used the `@property` annotation yet; this makes a lot of sense.

**Comment by kalessil** [comment]
You are welcome @cburschka )
